ext/mbstring/tests/gh10648.phpt

```diff
@@ -0,0 +1,29 @@
+--TEST--
+GH-10648 (mb_check_encoding() returns true for incorrect but interpretable ISO-2022-JP byte sequences)
+--EXTENSIONS--
+mbstring
+--FILE--
+<?php
+
+$jis_bytes = '1b244224221b2842'; // 'あ' in ISO-2022-JP
+$jis_bytes_without_esc = '1b24422422'; // 'あ' in ISO-2022-JP without trailing escape sequence
+
+$esc_kana = '1b2849211b2842'; // 'ｱ' in ISO-2022-JP (JIS X 0201 7bit kana with escape sequence)
+$so_kana = '0e210f'; // 'ｱ' in ISO-2022-JP (JIS X 0201 7bit kana with SO/SI)
+$gr_kana = 'b1'; // 'ｱ' in ISO-2022-JP (JIS X 0201 8bit kana)
+
+// Testing for trailing escape sequence
+var_dump(mb_check_encoding(hex2bin($jis_bytes), 'JIS'));
+var_dump(mb_check_encoding(hex2bin($jis_bytes_without_esc), 'JIS')); // false
+
+// Testing for JIS X 0201 kana
+var_dump(mb_check_encoding(hex2bin($esc_kana), 'JIS'));
+var_dump(mb_check_encoding(hex2bin($so_kana), 'JIS'));
+var_dump(mb_check_encoding(hex2bin($gr_kana), 'JIS')); // false
```

Comment:
Hmm, so `mb_check_encoding` will now treat GR kana as invalid even for `JIS`?? I thought this should be only for `ISO-2022-JP`.

What we call `JIS` should include both `JIS7` and `JIS8` variants of `ISO-2022-JP`, not so?

Reply:
I had mistakenly thought that JIS was a simple alias for ISO-2022-JP.
There is indeed a difference between the two in PHP.
https://3v4l.org/eFt7a
Then I think it is safe to say that ISO-2022-JP is strictly RFC1468 compliant and does not support JIS X 0201 kana in any way.
On the contrary, JIS supports JIS X 0201 kana in any way.
Specifically, the following code comments.
https://3v4l.org/Hb0lN
This behavior is different from any previous versions, but I think it makes sense.

Reply:
I found another difference between JIS and ISO-2022-JP.
JIS converts characters in the JIS X 0212 range, while ISO-2022-JP does not.

Here is a summary of how the three encodings ISO-2022-JP, JIS, and ISO-2022-JP-2004 convert or validate the three character sets JIS X 0208, JIS X 0212, and JIS X 0213.
https://3v4l.org/tSsN1
Since PHP 8.1, mb_check_encoding returns JIS X 0212 as valid as ISO-2022-JP, which needs to be fixed.

Unrelated to this issue, I think JIS should be converted to JIS X 0213 since there is no encoding called JIS-2004.

Reply:
Thanks so much. I still need to look at the comparison in 3v4l, haven't checked it out yet.